File: app/src/main/java/fr/xgouchet/chronorg/ui/presenters/BaseListPresenter.java
package fr.xgouchet.chronorg.ui.presenters;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.deezer.android.counsel.annotations.Trace;

import java.util.ArrayList;
import java.util.List;

import fr.xgouchet.chronorg.ui.fragments.BaseListView;
import rx.Observable;
import rx.Observer;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subscriptions.CompositeSubscription;

/**
 * @author Xavier Gouchet
 */
@Trace
public abstract class BaseListPresenter<T> implements ListPresenter<T> {

    @NonNull private final List<T> items;
    @NonNull private final CompositeSubscription subscriptions;

    @Nullable /*package*/ BaseListView<T> view;

    public BaseListPresenter() {
        items = new ArrayList<>();
        subscriptions = new CompositeSubscription();
    }

    public void setView(@NonNull BaseListView<T> view) {
        this.view = view;
        view.setPresenter(this);
    }


    @Override public void load(boolean force) {
        if (view == null) return;

        if (!force) {
            view.setContent(items);
        }

        view.setLoading(true);

        items.clear();

        Subscription subscription = getItemsObservable()
                .toList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<T>>() {
                    @Override public void onCompleted() {
                        view.setLoading(false);
                    }

                    @Override public void onError(Throwable e) {
                        view.setLoading(false);
                        view.setError(e);
                    }

                    @Override public void onNext(List<T> items) {
                        onItemsLoaded(items);
                    }
                });

        subscriptions.add(subscription);
    }

    protected abstract Observable<T> getItemsObservable();

    /*package*/ void onItemsLoaded(List<T> items) {
        this.items.addAll(items);
        if (view == null) return;

        if (this.items.isEmpty()) {
            view.setEmpty();
        } else {
            view.setContent(this.items);
        }
    }

    @Override public void itemSelected(@NonNull T item) {
        if (view == null) return;
        view.showItem(item);
    }

    @Override public void createNewItem() {
        if (view == null) return;
        view.showCreateItemUi();
    }

    @Override public void subscribe() {
        load(true);
    }

    @Override public void unsubscribe() {
        items.clear();
    }
}

File: khronorg/src/main/java/fr/xgouchet/khronorg/ui/presenters/BaseListPresenter.kt
package fr.xgouchet.khronorg.ui.presenters

import fr.xgouchet.khronorg.ui.navigators.Navigator
import fr.xgouchet.khronorg.ui.views.ListView
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.util.*

/**
 * @author Xavier F. Gouchet
 */
abstract class BaseListPresenter<T>(val navigator: Navigator<T>) : ListPresenter<T> {


    val items: MutableList<T> = ArrayList()

    var disposable: Disposable? = null

    var view: ListView<T>? = null


    override fun subscribe() {
        load(force = true)
    }

    override fun unsubscribe() {
        items.clear()
        disposable?.dispose()
    }

    override fun load(force: Boolean) {
        view?.let {

            if (!force) {
                it.setLoading(false)
                it.setContent(items)
                return@let
            }

            it.setLoading(true)
            items.clear()

            disposable?.dispose()
            disposable = getItemsObservable()
                    .toList()
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ onItemsLoaded(it) }, { onError(it) })
        }

    }

    override fun itemSelected(item: T?) {
        if (item == null) {
            navigator.goToItemCreation()
        } else {
            navigator.goToItemDetails(item)
        }
    }

    private fun onError(e: Throwable?) {
        view?.let {
            it.setLoading(false)
            if (e != null) it.setError(e)
        }
    }

    private fun onItemsLoaded(list: List<T>) {
        items.addAll(list)

        view?.let {
            it.setLoading(false)
            if (items.isEmpty()) {
                it.setEmpty()
            } else {
                it.setContent(items)
            }
        }
    }


    abstract fun getItemsObservable(): Observable<T>

}